### ISSUE TITLE
We should use official sqlite package

### ISSUE THREAD

**Issue by hol430** [APSIMInitiative/ApsimX#5317]
Is there any particular reason that we directly interact with the native SQLite functions? Seems like it would be more reliable (and simpler) to use the official sqlite packages from nuget.

**Comment by hol353** [comment]
Purely historical reasons probably. Can probably change to the Microsoft package.

**Comment by her123** [comment]
Just check that it will work with dotnetcore

**Comment by hol430** [comment]
Appears to be working fine on .net core as far as I can tell. Still trying to fix some issues related to closing the connection after closing a tab in the UI, but that's an issue with the .net framework version as well.

I wonder if it's worth keeping the database connection closed at all times except when we're actively using it. Might be a bit of work to implement though, so maybe it's not worth it...

**Comment by zur003** [comment]
Which of the SQLite nuget packages are you considering?

**Comment by zur003** [comment]
Probably wouldn't hurt to test this on Linux and OS X - it probably needs to load a native binary, and we need to be sure it works cross-platform.

**Comment by lie112** [comment]
I've used System.Data.SQLite.Core in other projects without trouble. 
I wasn't able to get it to work in a .Net Standard library but works in Core
I always find it best to keep SQL connections open to a minimum and always dispose of Commands otherwise the database is locked if you run the simulation again.
Looking forward to MAUI with .Net 5 and 6. That will solve some of the cross platform UI issues.

**Comment by zur003** [comment]
Thanks @lie112 - I wasn't aware of MAUI. It looks like it doesn't support Linux, though. @hol353 - do we need GUI support on Linux platforms?

**Comment by hol353** [comment]
Yes we need GUI support on Linux. MAUI is still a way off. We could spend some time evaluating Xamarin (which is where MAUI is evolving from) for our needs. Or we could try the MAUI preview later this year.

**Comment by hol430** [comment]
Yes, [System.Data.SQLite.Core](https://www.nuget.org/packages/Microsoft.Data.Sqlite.Core/) is the package I was considering. Haven't tested it on Linux yet but I will.

Migrating the UI to gtk3 is going to be a big job, but I expect that converting to another toolkit such as maui will be an even bigger job, so I don't really see much point unless it does something way better than gtk.

**Comment by her123** [comment]
I also found that the libsqlite3.so in the repository wouldn't work on Linux. I had to update it locally.

**Comment by hol430** [comment]
Where is your libsqlite3.so located? As far as I can see we don't have it anywhere in our repo

**Comment by her123** [comment]
It appeared in by Bin directory.

**Comment by her123** [comment]
It's in the repo

**Comment by hol430** [comment]
Ah, ok, it's there on mine too but I had deleted it. Do you recall why this was put in the Bin directory in the first place? I guess it should probably be removed since the documentation says sqlite needs to be installed separately anyway.

**Comment by her123** [comment]
It would be better to use the installed version (I think)

**Comment by zur003** [comment]
@hol430 - I don't see any reason to have this in the repository. The git log shows this dating back to 2013. I don't think we had a Linux installer at that point, so that may have been a reason for its inclusion. The installer is meant to take care of dependencies like sqlite, gtk2, mono, etc. Having an old version in the Bin directory is just asking for trouble.

**Comment by hol430** [comment]
Agreed. There are actually quite a number of files in the Bin directory which I really think shouldn't be there. In fact the repository probably shouldn't have a Bin directory at all. That azure stuff should be obtained from nuget (and the WindowsAzure stuff should be removed), the azure-apsim.exe doesn't get run by apsim and so doesn't need to be there, and the tools directory also doesn't need to be there.

**Comment by her123** [comment]
A cleanup would be good

**Comment by zur003** [comment]
@hol430 - You're probably the only one who understands the Azure stuff, and what is or isn't needed. A cleanup is overdue. Go for it!

**Comment by hol430** [comment]
Back on the sqlite side of things, I've tested on linux and it seems to have trouble pulling in the native interop layer out of the box. Specifically, I get this exception:

```
System.DllNotFoundException: SQLite.Interop.dll assembly:<unknown assembly> type:<unknown type> member:(null)
  at (wrapper managed-to-native) System.Data.SQLite.UnsafeNativeMethods.sqlite3_config_none(System.Data.SQLite.SQLiteConfigOpsEnum)
  at System.Data.SQLite.SQLite3.StaticIsInitialized () [0x0001d] in <237d5b0695c148fcaed7feddcce50827>:0 
  at System.Data.SQLite.SQLiteLog.Initialize () [0x00000] in <237d5b0695c148fcaed7feddcce50827>:0 
  at System.Data.SQLite.SQLiteConnection..ctor (System.String connectionString, System.Boolean parseViaFramework) [0x0002c] in <237d5b0695c148fcaed7feddcce50827>:0 
  at System.Data.SQLite.SQLiteConnection..ctor (System.String connectionString) [0x00000] in <237d5b0695c148fcaed7feddcce50827>:0 
  at (wrapper remoting-invoke-with-check) System.Data.SQLite.SQLiteConnection..ctor(string)
  at APSIM.Shared.Utilities.SQLite.OpenDatabase (System.String path, System.Boolean readOnly) [0x00001] in /home/drew/code/ApsimX2/APSIM.Shared/Utilities/SQLite.cs:68 
  at Models.Storage.DataStore.Open () [0x00037] in /home/drew/code/ApsimX2/Models/Storage/DataStore.cs:206 
  at Models.Storage.DataStore.OnCreated () [0x0000e] in /home/drew/code/ApsimX2/Models/Storage/DataStore.cs:169 
  at Models.Core.ApsimFile.FileFormat.ReadFromString[T] (System.String st, System.Collections.Generic.List`1[System.Exception]& creationExceptions, System.String fileName) [0x00098] in /home/drew/code/ApsimX2/Models/Core/ApsimFile/FileFormat.cs:105 
```

Nugt/msbuild seem to have correctly installed this file into my Bin directory, but mono seems to be having trouble loading it for some reason. If I compile it myself and copy the result into Bin/, then everything works fine. I expect we may encounter similar issues on mac as well.

**Comment by zur003** [comment]
@hol430 Is that with Mono or with .Net Core?

**Comment by hol430** [comment]
This is on mono, and there's no SQLite.Interop.dll.config file.

**Comment by zur003** [comment]
@hol430 - Ah, right. I wasn't thinking correctly about the config file. It appears that SQLite.Interop.dll is native, not managed. The .config files for managed assemblies contain dllmap sections that tell mono where to look for native components. I suppose to need to compile your own SQLite.Interop.dll (or equivalent .so or ,dylib) for non-Windows platforms....

**Comment by hol430** [comment]
Yeah, and that seems to work ok, with the manually-compiled .so file I can run simulations, look at graphs, etc. No idea if it will need to be recompiled again for Mac though - would you or @her123 be able to test it on Mac? I've got all the files up on a branch on my repo called "refactor/sqlite". I also included my notes on what/how I compiled the .so file, inside ApsimNG/Assemblies.

It's unfortunate that this is all necessary, I was hoping things would be simpler if we just use nuget for this.

**Comment by zur003** [comment]
That will almost certainly require recompilation for the Mac. We'll also need to check installers for those platforms to make certain the right stuff gets included.

**Comment by zur003** [comment]
@hol430 - It looks like the Microsoft.Data.Sqlite package on nuget might handle the cross-platform aspect better, but it appears to be designed for .Net Standard/Core, not Framework.

**Comment by hol430** [comment]
Oh, I didn't notice that one, that would certainly make things simpler. If it works out of the box on all platforms then it sounds like it might be easier to just continue using our current sqlite code for now and migrate to Microsoft.Data.SQLite at the same time as .net core. I need to get back onto that gtk3 stuff, I imagine the netcore branch must get stale pretty quickly..

**Comment by zur003** [comment]
@hol430 - I pulled your test code, and it works fine on Mac, but did require compilation of a new SQLite.Interop.dll (or rather libSQLite.Interop.dylib). We could use this, but as I mentioned before, we would need to be sure to package a suitable version of SQLite.Interop for each platform.

**Comment by hol430** [comment]
Thanks Eric. Would you be able to push up the .dylib file onto my branch? Unsure if we want to go ahead with this change or just wait for .net core and move to Microsoft.Data.SQLite, but at least this way we'll have the .dylib file handy if we do go ahead with System.Data.SQLite.

**Comment by zur003** [comment]
@hol430 - I couldn't work out how to push into your repository, but I've pushed the dylib onto my own sqlite branch. There is still much of git that I haven't mastered...

**Comment by hol430** [comment]
Thanks Eric - it's probably a problem with the permissions on my repo. I know I can push to some people's forks but not others.

**Comment by zur003** [comment]
@hol430 - you should now have a pull request for it. I put the file in the Bin folder, which is not a great place for it, but I didn't see an obvious alternative.

**Comment by hol430** [comment]
Merged - and yeah, I might end up making some DeploymentSupport sub-directories for Linux/mac at some point. We could put some of the other platform-specific files in there as well, like the .dll.config files.
